File: 프로그래머스/KaKao/이모티콘 할인행사.swift
// 2023 KAKAO BLIND RECRUITMENT 이모티콘 할인행사
// https://school.programmers.co.kr/learn/courses/30/lessons/150368
import Foundation

func permutation(_ arr: [Int], _ limit: Int) -> [[Int]] {
    var ret: [[Int]] = []
    
    func DFS(_ per: [Int], _ depth: Int) {
        if depth == limit {
            ret.append(per)
            return
        }
    
        for i in 0..<arr.count {
            DFS(per + [arr[i]], depth + 1)
        }
    }
    
    DFS([], 0)
    return ret
}

func getPlusServies(_ discount: [Int], _ users: [[Int]], _ emoticons: [Int]) -> (Int, Int) {
    var applyCount = 0
    var totalMoney = 0
    
    for user in users {
        var emoticonsTotal = 0
        
        for i in 0..<emoticons.count {
            if user[0] <= discount[i] {
                let prices = emoticons[i] - (emoticons[i] / 100 * discount[i])
                emoticonsTotal += prices
            }
            
            if user[1] <= emoticonsTotal {
                applyCount += 1
                emoticonsTotal = 0
                break
            }
        }
        
        totalMoney += emoticonsTotal
    }
    
    return (applyCount, totalMoney)
}

func solution(_ users:[[Int]], _ emoticons:[Int]) -> [Int] {
    var answer: [Int] = [0, 0]
    let discountPermutation = permutation([10, 20, 30, 40], emoticons.count)
    
    for discount in discountPermutation {
        let result: (applyCount: Int, totalMoney: Int) = getPlusServies(discount, users, emoticons)
        
        if  answer[0] <= result.applyCount {
            if answer[0] < result.applyCount {
                answer[0] = result.applyCount
                answer[1] = result.totalMoney
            } else {
                answer[1] = max(answer[1], result.totalMoney)
            }
        }
    }
    
    return answer
}

File: 프로그래머스/KaKao/이모티콘 할인행사.cpp
// https://school.programmers.co.kr/learn/courses/30/lessons/150368?language=cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int discount[4] = {10, 20, 30, 40};
vector<vector<int>> perArr;

// 중복순열
void backTracking(vector<int> per, int k, int level) {
    if (level == k) 
    {
        perArr.push_back(per);
        return;
    }
    
    for (int i = 0; i < 4; ++i)
    {
        per.push_back(discount[i]);
        backTracking(per, k, level+1);
        per.pop_back();
    }
    return;
}

vector<int> solution(vector<vector<int>> users, vector<int> emoticons) {
    vector<int> answer = {0, 0};
    backTracking({}, emoticons.size(), 0);
    
    for(auto& per: perArr)
    {
        int plusAply = 0;
        int sumMoney = 0;
        
        for (auto& user: users) 
        {
            int money = 0;
            for (int i = 0; i < emoticons.size(); ++i)
            {
                if (user[0] > per[i]) continue;
                money += emoticons[i] - emoticons[i]/100 * per[i];
            }
            
            if (money >= user[1])
            {
                plusAply++;
                money = 0;
            }
            sumMoney += money;
        }
        
        //cout << plusAply << ' ' << sumMoney << '\n';
        if (plusAply >= answer[0])
        {
            if (plusAply > answer[0])
            {
                answer[0] = plusAply;
                answer[1] = sumMoney;
            } else
            {
                answer[1] = max(answer[1], sumMoney);
            }
        }
    }
    
    return answer;
}